readthedocs/vcs_support/backends/git.py

```diff
@@ -68,6 +91,160 @@ def update(self):
+        Version.verbose_name = PR number (external versions)
+
+        :return: A refspec valid for fetch operation
+        """
+
+        if not self.version_type:
+            log.warning(
+                "Trying to resolve a remote reference without setting version_type is not "
+                "possible",
+                project_slug=self.project.slug,
+            )
+            return None
+
+        # Branches have the branch identifier set by the caller who instantiated the
+        # Git backend.
+        # If version_identifier is empty, then the fetch operation cannot know what to fetch
+        # and will fetch everything, in order to build what might be defined elsewhere
+        # as the "default branch". This can be the case for an initial build started BEFORE
+        # a webhook or sync versions task has concluded what the default branch is.
+        if self.version_type == BRANCH and self.version_identifier:
+            # Here we point directly to the remote branch name and update our local remote
+            # refspec to point here.
+            # The original motivation for putting 'refs/remotes/origin/<branch>' as the local refspec
+            # was an assumption in the .branches property that iterates over everything in
+            # refs/remotes/origin. We might still keep this pattern as it seems to be the most solid
+            # convention for storing a remote:local refspec mapping.
+            return (
+                f"refs/heads/{self.version_identifier}:refs/remotes/origin/"
+                f"{self.version_identifier}"
+            )
+        # Tags
+        if self.version_type == TAG and self.verbose_name:
+            # A "stable" tag is automatically created with Version.machine=True,
+            # denoting that it's not a branch/tag that really exists.
+            # Because we don't know if it originates from the default branch or some
+            # other tagged release, we will fetch the exact commit it points to.
+            if self.version_machine and self.verbose_name == "stable":
+                if self.version_identifier:
+                    return f"{self.version_identifier}:readthedocs/stable"
+                log.error("'stable' version without a commit hash.")
+                return None
+            return f"refs/tags/{self.verbose_name}:refs/tags/{self.verbose_name}"
+
+        if self.version_type == EXTERNAL:
+            # TODO: We should be able to resolve this without looking up in oauth registry
+            git_provider_name = self.project.git_provider_name
+
+            # Remote reference for Git providers where pull request builds are supported
+            if git_provider_name == GITHUB_BRAND:
+                return GITHUB_PR_PULL_PATTERN.format(id=self.verbose_name)
+            if self.project.git_provider_name == GITLAB_BRAND:
+                return GITLAB_MR_PULL_PATTERN.format(id=self.verbose_name)
+
+    def clone_ng(self):
```

Comment:
This needs a dosctring, not sure what the `_ng` prefix means here (maybe use a more explicit name?).

Reply:
sorry, yes, updated! been meaning to do that all the time :) needed a quick way to denote that this is for the "next version"... "ng" means "next generation", in this context it's meant with a ":wink:" - I think we'd probably want to rewrite a lot of stuff in the VCS backend :)